### DISCUSSION TITLE
pitch, yaw and T1,T2,R1,R2 using at.rotate_elem

### DISCUSSION THREAD

**Discussion by oscarxblanco** [atcollab/at#709] [Python AT]
Dear pyat team,

I've been checking the effect of pitch and yaw rotations on the T1,T2,R1and R2 definitions for an element in pyat and I see a behavior I do not expect.

For example, if I define a pitch over a drift
```python
import at
ring = at.Lattice([at.Drift('D1','0.3')],energy=3e9)
tilt0  = -0.0
pitch0 =  0.2
yaw0   =  0.0
at.rotate_elem(ring[0],tilt0,pitch0,yaw0)
print(f'{ring[0].T1=}')
print(f'{ring[0].T2=}')
print(f'{ring[0].R1=}')
print(f'{ring[0].R2=}')
```
I get
```bash
ring[0].T1=array([ 0.       ,  0.       , -0.0298004, -0.2      ,  0.       ,
        0.       ])
ring[0].T2=array([ 0.       ,  0.       , -0.0298004,  0.2      ,  0.       ,
        0.       ])
ring[0].R1=array([[ 1. ,  0. ,  0. ,  0. ,  0. ,  0. ],
       [ 0. ,  1. ,  0. ,  0. ,  0. ,  0. ],
       [ 0. ,  0. ,  1. ,  0. ,  0. ,  0. ],
       [ 0. ,  0. ,  0. ,  1. , -0.2,  0. ],
       [ 0. ,  0. ,  0. ,  0. ,  1. ,  0. ],
       [ 0. ,  0. ,  0. ,  0. ,  0. ,  1. ]])
ring[0].R2=array([[1. , 0. , 0. , 0. , 0. , 0. ],
       [0. , 1. , 0. , 0. , 0. , 0. ],
       [0. , 0. , 1. , 0. , 0. , 0. ],
       [0. , 0. , 0. , 1. , 0.2, 0. ],
       [0. , 0. , 0. , 0. , 1. , 0. ],
       [0. , 0. , 0. , 0. , 0. , 1. ]])

```
Which seems to be a small angle approximation with wrong signs, and missing components.

I expected the vertical momentum in T1 (i.e. T1[3], or the fourth component of T1) to be +tan(0.2). A positive pitch points downwards, toward negative y, and therefore, it is equivalent to a negative vertical momentum. A second negative comes from the definition of T1 because it is the opposite of the displacement introduced. The two negative signs (one from momentum and one from T1 definition) give a positive fourth component. Instead I get -0.2.

The longitudinal component in T1 and T2 can not be zero after a pitch rotation.

Also, angle components in T1,T2,R1 and R2 seems over simplified to small angle transformations where $\sin(x)\approx\tan(x)\approx x$.

The function also misses an option to rotate on the entrance instead of the center of the element, which , could be of help when defining angle error that do not change the entrance offsets.

**Comment by oscarxblanco** [comment]
I see that the help message of rotate_elem is missing yaw and pitch order.
```python
...
    The transformations are not all commmutative, the tilt is always the
    last transformation applied.
...
```

**Comment by oscarxblanco** [comment]
I have made a test on matlab using `SCgetTransformation.m` from Simulated Commissioning https://github.com/ThorstenHellert/SC. Although there is an asymmetry between T1 and T2 (probably due to some transformation I miss), this agrees more with what I expected when comparing to the sketched geometry of a pitch in terms of signs and overall values.

```bash
T1andT2fromSC =
         0         0   -0.0292    0.2027         0    0.0030
         0         0   -0.0304   -0.1987         0    0.0031
R1 =
    1.0000   -0.0031         0         0         0         0
         0    1.0000         0         0         0         0
         0         0    1.0203   -0.0031         0         0
         0         0         0    0.9801    0.1987         0
         0         0         0         0    1.0000         0
         0         0    0.2027   -0.0006         0    1.0000
R2 =
    1.0000    0.0031         0         0         0         0
         0    1.0000         0         0         0         0
         0         0    1.0203    0.0031         0         0
         0         0         0    0.9801   -0.1987         0
         0         0         0         0    1.0000         0
         0         0   -0.2027   -0.0006         0    1.0000
```

Here is the matlab script I used for test
```matlab
%% check rotations part 2
% oblanco 2023nov24
dx = 0;
dy = 0;
dz = 0;
atilt = 0.0;
apitch = 0.2;
ayaw = 0.0;
ax =  apitch;
ay =  ayaw;
az =  atilt;
magTheta = 0;
magLength = 0.3;
%[T1,T2,R1,R2] = SCgetTransformation(dx,dy,dz,ax,ay,az,magTheta,magLength,'refPoint','entrance');
[T1,T2,R1,R2] = SCgetTransformation(dx,dy,dz,ax,ay,az,magTheta,magLength,'refPoint','center');
T1andT2fromSC=[T1';T2']
R1, R2
```

**Comment by simoneliuzzo** [comment]
Dear @oscarxblanco,

I think we could update to the SC formulas. May be Lukas Malina (@lmalina) has already this converted to python in pySC https://github.com/lmalina/pySC . 
I have never used the function `rotate_elem` my self and seldom some homemade equivalent in matlab AT. Surely the matlab version I used was oversimplified as the studies where not solicited to have high precision (random errors rms).

**Comment by oscarxblanco** [comment]
Hi @simoneliuzzo , thank you for answering.
Alright. For most machines I believe that a small angle approximation is valid and would be enough.

My biggest concern with pyat is the sign disagreement wrt to the help message. You expect yaw and pitch angles to be defined as in the documentation but the code includes them with opposite sign, creating a mismatch with the position errors. This could create confusion when trying to extract conclusions from custom errors.

I would say that the pitch and yaw calculations done in `rotate_elem` should be fixed. The small angle approximation and the rotation order should be documented. And if possible, the option to rotate around the entrance point of the element should be implemented.

Apart from that, pySC could create another higher level functions to generalize rotations for higher precision, as is the case of SC.

**Comment by simoneliuzzo** [comment]
Dear @oscarxblanco,

no objection to the plan. I think you could open an issue on this.

best regards
Simone

**Comment by swhite2401** [comment]
Hello,

To continue the discussion I have a few comments on my side.

1-In the help it is mentioned that the rotations are performed clock-wise -> a positive pitch will result in the magnet pointing down, isn't this what we have now? If this is true this convention is counter-intuitive and should be changed in my opinion.
2-A pitch about the entrance is nothing but a pitch+ an offset. If we start mixing contributions it then becomes impossible to disentangle them and apply them as relative changes as is presently possible. It also introduces ambiguity when generating error tables with both pitch and offsets: in this case how do you define the offset? I would therefore strongly advocate to keep transformations independent, but this is up for discussion. A convenience function combining rotation and shift could be implemented to apply rotation about the entrance or exit if needed.
3-Concerning the order: pitch and yaw are commutable and can be applied in any order, what matters is that the tilt is always applied at the end. This allows to have reproducible behavior. I believe this is what is mentioned in the documentation but more details can be given if needed.

Concerning the small angle approximation this can easily be fixed if needed.

On the other hand, for the longitudinal shift I had some strong doubts that this is correct since we apply rotations by changing the particle coordinates and not moving the magnet itself or changing its effective length but maybe I am wrong... any thoughts on this points?

I agree with Simone that maybe this can be turned into an issue so that we do not forget to implement the requested changes.

**Comment by oscarxblanco** [comment]
Hi @swhite2401,
wrt 1-, the code is doing something different to the help message. A positive pitch should be seen as positive angle in T1 and negative angle in T2. If I use pitch=0.2 the code does the opposite 
```bash
ring[0].T1=array([ 0.       ,  0.       , -0.0298004, **-0.2**wrongsign**      ,  0.       ,
        0.       ])
ring[0].T2=array([ 0.       ,  0.       , -0.0298004,  **0.2**wrongsign**      ,  0.       ,
        0.       ])
```
There's a similar problem in yaw definition.

**Reply by swhite2401** [reply]
Ah yes! Too many signs, I am getting confused.

**Comment by oscarxblanco** [comment]
@swhite2401 
wrt 2-, yes, I agree, it is a combination of rotations and offsets. I suggested the possibility to change the pivot when doing the transformation because many other element parameters refer to the entrance. T1 would therefore be the negative of whatever dx,dy,dz is defined, and the rotation effect will be only in T2.

However, it could be enough to include somewhere in the doc string that rotations are applied at the center of the element, so, it is clear that T1 and T2 offsets would correspond to displacements + or - other component from rotation.

**Comment by oscarxblanco** [comment]
@swhite2401 
wrt 3-, pitch and yaw rotations commute, but, pitch-yaw-tilt don't. The function is written to define any combination of the three components and should therefore document the order in which rotations are applied.

The doc message says
```
    The transformations are not all commmutative, the tilt is always the
    last transformation applied.
```
which is correct, but, leaves space to doubt with respect to what the code does.

**Comment by oscarxblanco** [comment]
@swhite2401 
wrt to the longitudinal effect of a rotation. It effectively increases the length by a factor $\sec(\theta)-1$. A simple image would be to consider the difference between crossing a square magnet through the center on one of the sides, and crossing through any of the diagonals in which case the magnet length increases geometrically by a factor $\sqrt(2)-1$ (which is exactly $\sec(45/360\times2\pi)-1$).

It might be a second order effect depending on the magnet length and rotation. Therefore, if the small angle approximation is implemented it would be worth saying it in the doc, stating that only transverse angles and positions are modified and other effects are ignored.

**Reply by swhite2401** [reply]
Sure but adding a `ct` shift as it seems to be the case in SC is incorrect and does not solve the issue you describe. I would rather keep the present approximation and document it properly instead of adding something that give the impression that the effective magnet length is taken into account while it is not.

**Comment by swhite2401** [comment]
Ok thanks @oscarxblanco for all these inputs, clearly the documentation needs to be significantly improved!

Concerning the small angle approximation, do you think it is worth the effort or is the error negligible and we can live with the present implementation?

**Comment by oscarxblanco** [comment]
@swhite2401 
In general, there are a few more transformations to make the rotation implementation completely valid for tracking. Thorsten Heller's `SCgetTransformation.m` opts for an intermediate step of including a wedge drift at the magnet entrance. I believe that his geometrical approximation was compared to PTC which chooses an analytical method to connect transport maps.

The rotation function `at.rotate_elem` is limited to a small angle approximation on the transverse planes and discards any other effects. From my point of view this is perfectly valid, but, it should be clearly stated in the documentation.

**Reply by swhite2401** [reply]
Ok I will do
